### PR TITLE
Registries for extending compose-rl

### PR DESCRIPTION
**Pull request by gupta-abhay**
- As name suggests
- Enables private development on top of existing compose-rl repo
- Current registries are for kl_controllers and rewards (we use existing dataloder registries from foundry)

### CODE REVIEW

File: compose_rl/ppo/reward_manager.py [hunk 1]
```diff
@@ -104,13 +104,18 @@ def __init__(
...
 
             if issubclass(reward_cls, Reward):
                 # TODO: This assumes that all functional rewards are document level rewards.
                 # This is not necessarily true, but is a reasonable assumption for now.
                 self.granularities[reward_name] = 'document'
-                model = reward_cls(reward_config, self.tokenizer)
+                model = build_rewards(
+                    name=reward_type,
+                    tokenizer=self.tokenizer,
+                    kwargs={'cfg': reward_config},
```
Comment (dakinggg) [inline compose_rl/ppo/reward_manager.py:117]: it would be nicer to flatten the 'cfg', so that the reward classes don't take in one dictionary arg called `cfg`, but instead take in named args. so it would be `kwargs=reward_config` here.

Comment (dakinggg) [reply compose_rl/ppo/reward_manager.py:117]: and same in other places

File: compose_rl/ppo/kl_controller.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ class AdaptiveKLController(BaseKLController):
     """
 
     def __init__(self, kl_config: dict, device: str = 'cpu'):
-        super().__init__(device=device)
+        super().__init__(kl_config=kl_config, device=device)
```
Comment (dakinggg) [inline compose_rl/ppo/kl_controller.py:51]: like this could just take in `init_kl_coef`, `target`, and `horizon`, as named args rather than an arbitrary dict

File: compose_rl/registry.py [hunk 3]
```diff
@@ -1,43 +1,58 @@
...
+    description=_kl_controller_description,
+)
+
+
+def initialize_registries():
+    """Initialize registries with their types after all modules are loaded."""
+    from compose_rl.ppo.kl_controller import BaseKLController
+    from compose_rl.reward_learning.base_reward import BaseReward
+
+    rewards.generic_type = type[BaseReward]  # type: ignore
```
Comment (dakinggg) [inline compose_rl/registry.py:54]: I don't think this really does anything cause the only point of the generic types are for static typing. The way I would do this is to create in `interfaces` folder in `compose_rl`, and move the `BaseKLController` and `BaseReward` there. The interfaces folder should not import from anywhere else in the library, to avoid circular imports.

File: compose_rl/registry_builders.py [hunk 4]
```diff
@@ -1,19 +1,88 @@
...
-        kl_ctl_type = 'fixed'
-        cfg['kl_ctl_type'] = kl_ctl_type
-        cfg['init_kl_coef'] = 0.05
+log = logging.getLogger(__name__)
 
-    return KL_CONTROLLER_REGISTRY[kl_ctl_type](cfg)
+Tokenizer = Optional[Union[PreTrainedTokenizer, PreTrainedTokenizerFast]]
+
+
+def build_kl_controllers(
```
Comment (dakinggg) [inline compose_rl/registry_builders.py:25]: super nit, i think the builders should be singular (`build_kl_controller`), since when you call it, you are only building one.

File: compose_rl/registry_builders.py [hunk 5]
```diff
@@ -1,19 +1,88 @@
...
+    name: str,
+    kl_config: dict[Any, Any],
```
Comment (dakinggg) [inline compose_rl/registry_builders.py:27]: its probably [str, Any]?

File: compose_rl/registry_builders.py [hunk 6]
```diff
@@ -1,19 +1,88 @@
...
+    Args:
+        name (str): Name of the reward model to build.
+        tokenizer: Tokenizer: The tokenizer to use for the reward model.
+        kwargs: dict[Any, Any]: Additional keyword arguments.
+
+    Returns:
+        BaseReward: The reward model.
+    """
+    registry_to_use = registry.rewards
+    if name in registry.rewards:
```
Comment (dakinggg) [inline compose_rl/registry_builders.py:71]: not sure i quite understand the goal of this if statement. `registry_to_use` is always `registry.rewards`?

File: compose_rl/reward_learning/base_reward.py [hunk 7]
```diff
@@ -20,7 +20,9 @@ class BaseReward(ABC):
             will reference this attribute.
 
     Args:
+        cfg: the config for the reward class.
```
Comment (dakinggg) [inline compose_rl/reward_learning/base_reward.py:23]: same here with the cfg. would rather use named args in the subclasses.

File: compose_rl/ppo/kl_controller.py (hunk 2 above)
Comment (gupta-abhay) [reply compose_rl/ppo/kl_controller.py:18]: should be doable for kl_controller - but for rewards, its tricky. There can be many rewards we can define and each can have different variables they need to execute it. Will fix for the controllers - lmk if that sounds good.

Comment (gupta-abhay) [reply compose_rl/ppo/kl_controller.py:18]: I agree - but then diff classes use diff vars and sometimes they are used as floats or other times they are defined as tensors -- and then having to make them optional and then deal with pyright is :(

Comment (gupta-abhay) [reply compose_rl/ppo/kl_controller.py:18]: same with rewards - diff rewards require diff values and those are not all shared by diff classes, leading to a lot of manual handling of optional variables.

File: compose_rl/ppo/kl_controller.py [hunk 8]
```diff
@@ -47,15 +13,25 @@ class AdaptiveKLController(BaseKLController):
     KL = clip((kl_curr / target) - 1, -n, n)
     """
 
-    def __init__(self, kl_config: dict, device: str = 'cpu'):
-        super().__init__(device=device)
-        self._value = kl_config['init_kl_coef']
-        self._target = kl_config['target']
-        self._horizon = kl_config['horizon']
+    def __init__(
```
Comment (dakinggg) [inline compose_rl/ppo/kl_controller.py:16]: can add docstring now that these args are defined?

File: compose_rl/ppo/kl_controller.py [hunk 9]
```diff
@@ -47,15 +13,25 @@ class AdaptiveKLController(BaseKLController):
...
+    ):
+        self._value = init_kl_coef
+        self._target = target
+        self._horizon = horizon
+        self.device = device
 
     def update(self, current: torch.Tensor, n_steps: int):
         target = self._target
-        proportional_error = np.clip(current / target - 1, -0.2, 0.2)
+        proportional_error = torch.clamp(
```
Comment (dakinggg) [inline compose_rl/ppo/kl_controller.py:30]: did you intend to include this change? i mean, seems fine, jw

File: compose_rl/ppo/kl_controller.py [hunk 10]
```diff
@@ -75,9 +51,9 @@ def load_state_dict(self, state_dict: dict[str, float]):
 class FixedKLController(BaseKLController):
     """Fixed KL controller that always returns same value."""
 
-    def __init__(self, kl_config: dict, device: str = 'cpu'):
-        super().__init__(device=device)
-        self._value = kl_config['init_kl_coef']
+    def __init__(self, init_kl_coef: float, device: str = 'cpu'):
```
Comment (dakinggg) [inline compose_rl/ppo/kl_controller.py:54]: same, docstring if you want

File: compose_rl/ppo/kl_controller.py [hunk 11]
```diff
@@ -99,15 +75,23 @@ class KLPIDController(BaseKLController):
     KL = KL + -kl_lr * ((kl_curr - target) * trajectory_len)
     """
 
-    def __init__(self, kl_config: dict, device: str = 'cpu'):
+    def __init__(
+        self,
+        init_kl_coef: float,
```
Comment (dakinggg) [inline compose_rl/ppo/kl_controller.py:80]: same

File: compose_rl/ppo/kl_controller.py [hunk 12]
```diff
@@ -144,15 +128,23 @@ class BallKLController(BaseKLController):
     KL = clamp(KL + -kl_lr * ((kl_curr - target) * trajectory_len))
     """
 
-    def __init__(self, kl_config: dict, device: str = 'cpu'):
+    def __init__(
+        self,
+        init_kl_coef: float,
```
Comment (dakinggg) [inline compose_rl/ppo/kl_controller.py:133]: same

File: compose_rl/registry_builders.py [hunk 13]
```diff
@@ -1,19 +1,87 @@
...
+    if 'tokenizer' in kwargs:
+        raise ValueError(
+            f'`tokenizer` is a reserved keyword for rewards. Please remove it from the kwargs.',
+        )
+    kwargs['tokenizer'] = copy.deepcopy(tokenizer)
+
+    return construct_from_registry(
+        name=name,
+        registry=registry.rewards,
+        partial_function=True,
```
Comment (dakinggg) [inline compose_rl/registry_builders.py:83]: you can remove this arg here, `True` is the default, and it has no impact anyway since you are registering classes.

File: compose_rl/reward_learning/functional.py [hunk 14]
```diff
@@ -26,8 +26,9 @@ class IncreasingNumbersReward(Reward):
     # This can be run async
     BLOCKING = False
 
-    def __init__(self, cfg: dict[Any, Any], tokenizer: Tokenizer):
-        super().__init__(cfg, tokenizer)
+    def __init__(self, reward: float, tokenizer: Tokenizer):
```
Comment (dakinggg) [inline compose_rl/reward_learning/functional.py:29]: is `reward` unused? i think yes?

File: compose_rl/reward_learning/functional.py [hunk 15]
```diff
@@ -248,9 +234,10 @@ class BaseVerifierReward(Reward):
     # This can be run async
     BLOCKING = False
 
-    def __init__(self, cfg: dict[Any, Any], tokenizer: Tokenizer):
-        super().__init__(cfg, tokenizer)
-        self.reward = cfg.get('reward', 1.0)
+    def __init__(self, reward: float, tokenizer: Tokenizer):
+        super().__init__(tokenizer)
+        self.reward = reward
```
Comment (dakinggg) [inline compose_rl/reward_learning/functional.py:239]: did you want to keep the previous impl's default of 1.0?

File: compose_rl/reward_learning/inference_model.py [hunk 16]
```diff
@@ -23,16 +23,20 @@ class InferenceRewardModel(RewardModel):
...
-    def __init__(self, cfg: dict[Any, Any], tokenizer: Tokenizer):
-        super().__init__(cfg, tokenizer)
-
-        deployment_str: str = self.cfg.get('deployment', None)
-        assert deployment_str is not None, 'InferenceRewardModel requires a deployment identifier.'
-        self.deployment_name = get_remote_name(deployment_str)
+    def __init__(
+        self,
+        deployment: str,
+        max_retries: int,
```
Comment (dakinggg) [inline compose_rl/reward_learning/inference_model.py:29]: did you want to keep the rpevious default of 5?

File: compose_rl/reward_learning/inference_model.py [hunk 17]
```diff
@@ -23,16 +23,20 @@ class InferenceRewardModel(RewardModel):
...
+        timeout: Optional[int],
```
Comment (dakinggg) [inline compose_rl/reward_learning/inference_model.py:30]: same, keep previous defaults of `None` for timeout and threshold?

**Review by dakinggg — APPROVED** [review]
LGTM, some nits, and qs if you wanted to keep previous default value behavior

File: compose_rl/ppo/callback.py [hunk 18]
```diff
@@ -394,7 +394,13 @@ def __init__(
         # Programmatically setting the max buffer size instead of the yaml
         var_config['buffer']['max_buffer_size'] = self.num_batches_per_update
         self.buffer = MinibatchRolloutBuffer(var_config['buffer'])
-        self.kl_ctl = build_kl_controller(var_config['kl_controller'])
+
+        # Build the KL controller through registries
+        kl_ctl_name = var_config['kl_controller'].pop('kl_ctl_type')
```
Comment (vchiley) [inline compose_rl/ppo/callback.py:399]: Should we enable kl controller to be None
for no KL penalty?

File: compose_rl/interfaces/base_reward.py [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    BLOCKING: bool = True
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        self.tokenizer = tokenizer
+        pass
```
Comment (vchiley) [inline compose_rl/interfaces/base_reward.py:40]:
```suggestion
        del args, kwargs  # unsued here
        pass
```

File: compose_rl/interfaces/base_kl_controller.py [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    """Abstract base class for KL controller classes.
+
+    Args:
+        device: the device to run the KL controller on.
+        *args: additional arguments.
+        **kwargs: additional keyword arguments.
+    """
+
+    def __init__(self, device: str, *args: Any, **kwargs: Any):
+        del device, args, kwargs  # unsued here
```
Comment (vchiley) [inline compose_rl/interfaces/base_kl_controller.py:22]:
```suggestion
        del args, kwargs  # unsued here
        self.device = device
```

why not set device under the `super()` call

File: compose_rl/ppo/callback.py (hunk 18 above)
Comment (gupta-abhay) [reply compose_rl/ppo/callback.py:399]: is a meaningful ask - but better handled in another PR - touches several aspects of the code.
